packages/flutter/lib/src/material/navigation_drawer.dart

```diff
@@ -0,0 +1,416 @@
+///  * <https://material.io/design/components/navigation-drawer.html>
+class NavigationDrawer extends StatelessWidget {
+  /// Creates a Material 3 Navigation Drawer component.
+  const NavigationDrawer({
+    super.key,
+    required this.children,
+    this.backgroundColor,
+    this.surfaceTintColor,
+    this.elevation,
+    this.onDestinationSelected,
+    this.selectedIndex = 0,
+  });
+
+  /// Sets the color of the [Material] that holds all of the [NavigationDrawer]'s
+  /// contents.
+  ///
+  /// If this is null, then [NavigationDrawerThemeData.backgroundColor] is used.
+  /// If that is also null, then it falls back to [Material]'s default.
+  final Color? backgroundColor;
+
+  /// Sets the color of the [Material] that holds all of the [NavigationDrawer]'s
+  /// contents.
+  ///
+  /// If this is null, then [NavigationDrawerThemeData.surfaceTintColor] is used.
+  /// If that is also null, then it falls back to [Material]'s default.
+  final Color? surfaceTintColor;
+
+  /// The elevation of the [NavigationDrawer] itself.
+  ///
+  /// If null, [NavigationDrawerThemeData.elevation] is used. If that
+  /// is also null, it will be 1.0.
+  final double? elevation;
+
+  /// Defines the appearance of the items within the navigation drawer.
+  ///
+  /// The list contains [NavigationDrawerDestination] widgets and/or customized
+  /// widgets like headlines and dividers.
+  final List<Widget> children;
+
+  /// The index into destinations for the current selected
+  /// [NavigationDrawerDestination] or null if no destination is selected.
+  ///
+  /// Only [NavigationDrawerDestination] is counted in index.
+  final int? selectedIndex;
+
+  /// Called when one of the destinations is selected.
+  ///
+  /// This callback usually updates the int passed to [selectedIndex].
+  ///
+  /// Upon updating [selectedIndex], the [NavigationDrawer] will be rebuilt.
+  final ValueChanged<int>? onDestinationSelected;
+
+  @override
+  Widget build(BuildContext context) {
+    final NavigationDrawerThemeData navigationDrawerTheme =
+        NavigationDrawerTheme.of(context);
+    final NavigationDrawerThemeData defaults =
+        _NavigationDrawerDefaultsM3(context);
+    final int totalNumberOfDestinations =
+        children.whereType<NavigationDrawerDestination>().toList().length;
+
+    final List<Widget> newChildren = <Widget>[];
+    for (int i = 0, j = 0; i < children.length; i++) {
+      final Widget e = children[i];
+      if (e is! NavigationDrawerDestination) {
+        newChildren.add(e);
+      } else {
+        final int index = j;
+        newChildren.add(
+          SelectableAnimatedBuilder(
+            duration: const Duration(milliseconds: 500),
+            isSelected: index == selectedIndex,
+            builder: (BuildContext context, Animation<double> animation) {
+              return NavigationDestinationInfo(
+                index: index,
+                totalNumberOfDestinations: totalNumberOfDestinations,
```

Comment:
Why is the total number of destinations needed for the child? If it really isn't then we don't need to traverse the children list twice during the build method.

Reply:
total number of destinations  is used for semantics to announce something like "1 of 5"